File: lab_1/src/KMPMatcher.cpp
#include "KMPMatcher.hpp"

KMPMatcher::KMPMatcher(const std::string& pattern)
    : pattern_(pattern), startingState_(0) {
    // Precompute prefix function for the given pattern so runSearch can be const
    computePrefixFunction();
}

void KMPMatcher::computePrefixFunction() {
    prefixFunction_.clear();
    const int n = static_cast<int>(pattern_.size());
    prefixFunction_.resize(n);
    if (n == 0) return;

    prefixFunction_[0] = 0;
    for (int i = 1; i < n; ++i) {
        int j = prefixFunction_[i - 1];
        while (j > 0 && pattern_[i] != pattern_[j]) {
            j = prefixFunction_[j - 1];
        }
        if (pattern_[i] == pattern_[j]) ++j;
        prefixFunction_[i] = j;
    }
}

std::vector<int> KMPMatcher::runSearch(const std::string& text) const {
    std::vector<int> matches;
    const int m = static_cast<int>(pattern_.size());
    const int n = static_cast<int>(text.size());
    if (m == 0 || n == 0 || n < m) return matches;

    // prefixFunction_ should have been computed in constructor; guard just in case
    if (static_cast<int>(prefixFunction_.size()) != m) return matches;

    int j = 0; // number of chars matched
    for (int i = 0; i < n; ++i) {
        while (j > 0 && text[i] != pattern_[j]) {
            j = prefixFunction_[j - 1];
        }
        if (text[i] == pattern_[j]) ++j;
        if (j == m) {
            matches.push_back(i - m + 1);
            j = prefixFunction_[j - 1];
        }
    }
    return matches;
}

void KMPMatcher::reset() {
    // Clear computed structures; keep pattern_ intact
    automateStates_.clear();
    acceptingStates_.clear();
    alphabet_.clear();
    prefixFunction_.clear();
    startingState_ = 0;
}

File: lab_1_python/src/kmp_matcher.py
from typing import List

class KMPMatcher:
    def __init__(self, pattern: str):
        self.pattern = pattern
        self.prefix_function: List[int] = []

    def compute_prefix_function(self) -> None:
        m = len(self.pattern)
        self.prefix_function = [0] * m
        k = 0
        for q in range(1, m):
            while k > 0 and self.pattern[k] != self.pattern[q]:
                k = self.prefix_function[k - 1]
            if self.pattern[k] == self.pattern[q]:
                k += 1
            self.prefix_function[q] = k

    def run_search(self, text: str) -> List[int]:
        if not self.prefix_function:
            self.compute_prefix_function()
        n = len(text)
        m = len(self.pattern)
        result: List[int] = []
        q = 0
        for i in range(n):
            while q > 0 and (q >= m or self.pattern[q] != text[i]):
                q = self.prefix_function[q - 1]
            if q < m and self.pattern[q] == text[i]:
                q += 1
            if q == m:
                result.append(i + 1 - m)
                q = self.prefix_function[q - 1]
        return result

    def reset(self) -> None:
        self.prefix_function = []
